packages/server-wallet/src/onchain-service/__chain-test__/transaction-submission.test.ts

```diff
@@ -0,0 +1,111 @@
+import {Wallet, providers, BigNumber} from 'ethers';
+import {randomChannelId} from '@statechannels/nitro-protocol';
+
+import {TransactionSubmissionService, TransactionSubmissionError} from '../transaction-submission';
+import config from '../../config';
+import {TransactionSubmissionStore} from '../store';
+
+import {MinimalTransaction} from '..';
+
+const verifyResponse = async (
+  response: providers.TransactionResponse,
+  tx: MinimalTransaction
+): Promise<void> => {
+  expect(response.to).toEqual(tx.to);
+  expect(response.value?.toString()).toEqual(tx.value?.toString());
+  expect(response.data).toEqual(tx.data);
+  expect(response.hash).toBeDefined();
+  const receipt = await response.wait();
+  expect(receipt.transactionHash).toBe(response.hash);
+  expect(receipt.to).toBe(response.to);
+};
+
+const getTransaction = (overrides: Partial<MinimalTransaction> = {}): MinimalTransaction => {
+  return {
+    to: Wallet.createRandom().address,
+    value: BigNumber.from(10),
+    data: '0x00',
+    ...overrides,
+  };
+};
+
+const getFailingWallet = (error?: string): any => {
+  return {
+    connect: (_provider: providers.JsonRpcProvider): any => {
+      return {
+        getTransactionCount: (): number => 0,
+        sendTransaction: (
+          _tx: providers.TransactionRequest
+        ): Promise<providers.TransactionResponse> => {
+          throw new Error(error || 'Fail');
+        },
+      };
+    },
+  };
+};
+
+describe('TransactionSubmissionService.submitTransaction', () => {
+  let service: TransactionSubmissionService;
+  let wallet: Wallet;
+  let provider: providers.JsonRpcProvider;
+
+  const channelId = randomChannelId();
+
+  beforeEach(async () => {
+    provider = new providers.JsonRpcProvider(config.rpcEndpoint);
+    wallet = new Wallet(config.serverPrivateKey);
+    service = new TransactionSubmissionService(provider, wallet, new TransactionSubmissionStore());
+  });
+
+  it('should fail if provided with 0 attempts', async () => {
+    const tx = getTransaction();
+    await expect(service.submitTransaction(channelId, tx, {maxSendAttempts: 0})).rejects.toEqual(
+      new TransactionSubmissionError(TransactionSubmissionError.reasons.zeroAttempts)
+    );
+  });
+
+  it('should send a transaction successfully', async () => {
+    const tx = getTransaction();
+    const response = await service.submitTransaction(channelId, tx);
+    await verifyResponse(response, tx);
+  });
+
+  it('should send concurrent transactions successfully', async () => {
+    const tx = getTransaction();
+    const responses = await Promise.all([
+      service.submitTransaction(channelId, tx),
+      service.submitTransaction(channelId, tx),
+    ]);
+    await Promise.all(responses.map(response => verifyResponse(response, tx)));
+  });
+
+  it('should retry transactions with known errors', async () => {
+    // Create a "wallet" that forces an error
+    const mock = getFailingWallet(TransactionSubmissionError.knownErrors.badNonce);
```

Comment:
I like the `knownErrors` pattern, but it might deserve a better name and place.

When I first saw it, I was confused, thinking that what you were calling `knownErrors` is what we call `reasons`: the types of errors that this thing is expected to **throw**.

Instead, they are the types of errors that this thing internally **catches**, and that is a bit of an implementation detail. Therefore, `knownErrors` doesn't seem like it belongs to the `TransactionSubmissionError` class.

I [can't think](https://martinfowler.com/bliki/TwoHardThings.html) of a better name off the top of my head. Something like `EthersErrorsThatWeHandle` is not good.